### PR TITLE
feat: add `config.init` method [MX-1122]

### PR DESCRIPTION
**Pull request by Alex-Joseph**
https://rewardops.atlassian.net/browse/MX-1122

- created `config.init`
- added config validation
- added warning when calling `config.set`
- removed as many instances of `config.set` as possible

### CODE REVIEW

File: lib/schemas/config.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+const yup = require('yup');
+
+const configSchema = yup.object().shape({
```
Comment (Alex-Joseph) [inline lib/schemas/config.js:3]: still working on the tests for this...

Comment (Alex-Joseph) [reply lib/schemas/config.js:3]: done: https://github.com/rewardops/rewardops-sdk-node/pull/46/files#diff-dab480eedb2fe55653dda98093b822ab

File: lib/config.js [hunk 2]
```diff
@@ -63,20 +64,54 @@ const getAll = () => cloneDeep(config);
  * @param {*} value Value of a configuration property
  *
  * @returns {*} The value given.
+ *
+ * @deprecated since version 1.3
  */
 const set = (key, value) => {
-  config[key] = value;
+  // NOTE: we cannot use `log` function here as this module is a dependency of logger module.
+  console.warn('config.set is deprecated, please use config.init instead');
```
Comment (shamrt) [inline lib/config.js:72]: I think this is okay for now, but seeing as the developer-user may not be using a console, we'll want to consider a different approach for v2

File: test/schemas/store-order-recipient.test.js [hunk 3]
```diff
@@ -2,13 +2,15 @@ const faker = require('faker');
...
 const { storeOrderRecipientSchema, isNumberOrUuid } = require('../../lib/schemas/store-order-recipient');
+const { mockConfig } = require('../test-helpers/mock-config');
 
 const testLocale = faker.random.locale();
 const testNumber = faker.random.number();
 
 describe('storeOrderRecipientSchema.validate()', () => {
   beforeEach(() => {
-    config.set('supportedLocales', [testLocale]);
+    config.reset();
```
Comment (shamrt) [inline test/schemas/store-order-recipient.test.js:12]: I see this is a shorthand, but for safety, `config.reset()` should go into a `afterEach`

File: test/test-helpers/mock-config.js [hunk 4]
```diff
@@ -21,7 +21,8 @@ const mockConfig = ({
   timeout = 20000,
   verbose = true,
   supportedLocales = undefined,
-}) => ({
+  logFilePath = faker.system.filePath(),
```
Comment (shamrt) [inline test/test-helpers/mock-config.js:24]: 🚀

File: test/urls.test.js [hunk 5]
```diff
@@ -16,7 +17,7 @@ describe('urls', () => {
 
   describe('getApiServerUrl()', () => {
     afterAll(() => {
```
Comment (shamrt) [inline test/urls.test.js:19]: Not changes here, but `afterAll` should be `afterEach`

Comment (shamrt) [reply test/urls.test.js:19]: In fact, this `afterAll` can be omitted and the setup methods in the parent scope should be `*Each`s

File: test/urls.test.js [hunk 6]
```diff
@@ -72,26 +73,24 @@ describe('urls', () => {
...
-
-      config.set('apiVersion', 'v6-beta');
-
-      expect(urls.getApiBaseUrl()).toEqual(`${urls.getApiServerUrl()}/api/v6-beta`);
+    test.each`
+      apiVersion | expectedPath
+      ${'v3'}    | ${'/api/v3'}
+      ${'v5'}    | ${'/api/v5'}
+    `('given $apiVersion, $expectedPath is present in the url', ({ apiVersion, expectedPath }) => {
+      config.reset();
```
Comment (shamrt) [inline test/urls.test.js:90]: Then we can
```suggestion
```

File: test/utils/axios-helper.test.js [hunk 7]
```diff
@@ -6,16 +6,21 @@ const config = require('../../lib/config');
 const auth = require('../../lib/auth');
 const { setPiiToken } = require('../../lib/utils/axios-helpers');
 const { getTestAccessToken } = require('../test-helpers/auth-helpers');
+const { mockConfig } = require('../test-helpers/mock-config');
 
 const RoUrl = faker.internet.url();
 const piiUrl = faker.internet.url();
```
Comment (shamrt) [inline test/utils/axios-helper.test.js:12]: Nit (optional): May as well use the same approach in L11–12 as in L13–14
```suggestion
const apiServerUrl = faker.internet.url();
const piiServerUrl = faker.internet.url();
```

File: test/utils/axios-helper.test.js [hunk 8]
```diff
@@ -6,16 +6,21 @@ const config = require('../../lib/config');
...
 const clientSecret = faker.internet.password();
 
-config.set('apiServerUrl', RoUrl);
-config.set('piiServerUrl', piiUrl);
-config.set('clientId', clientId);
-config.set('clientSecret', clientSecret);
+config.init(
+  mockConfig({
+    apiServerUrl: RoUrl,
+    piiServerUrl: piiUrl,
```
Comment (shamrt) [inline test/utils/axios-helper.test.js:19]: Then we can
```suggestion
    apiServerUrl,
    piiServerUrl,
```

File: test/config.test.js [hunk 9]
```diff
@@ -1,6 +1,48 @@
...
+const validConfig = {
+  apiServerUrl: faker.internet.url(),
+  apiVersion: faker.random.arrayElement(['v3', 'v4', 'v5']),
+  piiServerUrl: faker.internet.url(),
+  clientId: getTestAccessToken(),
+  clientSecret: getTestAccessToken(),
+};
 
 describe('config', () => {
+  describe('init()', () => {
```
Comment (shamrt) [inline test/config.test.js:15]: Nit: A convention when referencing a module or class method is to preface its name with a `#`:
```suggestion
  describe('#init', () => {
```
We're doing the same elsewhere in the library (see https://github.com/rewardops/rewardops-sdk-node/pull/46/files#diff-61b19a774848afe700b799296ff39a82R16), so I think we should continue that for consistency.

File: test/config.test.js [hunk 10]
```diff
@@ -1,6 +1,48 @@
...
+      expect(throwsValidationError).toThrow(
+        expect.objectContaining({
+          name: 'ValidationError',
+          message: 'apiVersion must be one of the following values: v3, v4, v5',
+        })
+      );
+    });
+  });
+
   describe('getAll()', () => {
```
Comment (shamrt) [inline test/config.test.js:46]: So same here:
```suggestion
  describe('#getAll', () => {
```

Comment (shamrt) [reply test/config.test.js:46]: And below

File: test/fixtures/v3/program-orders.fixtures.js [hunk 11]
```diff
@@ -3,8 +3,6 @@ const nock = require('nock');
 const RO = require('../../..');
 const { generateBasicAuthToken } = require('../../../lib/utils/auth');
 
-RO.config.set('apiVersion', 'v3');
```
Comment (shamrt) [inline test/fixtures/v3/program-orders.fixtures.js:6]: I suppose we should deprecate `v3` entirely, given we've done so in Pangea. Can you spool up a ticket for the modernization Epic, please?

File: test/resources/v3/orders.test.js [hunk 12]
```diff
@@ -1,9 +1,9 @@
 const RO = require('../../..');
+const { mockConfig } = require('../../test-helpers/mock-config');
 
 describe('v3 orders', () => {
   beforeAll(() => {
-    RO.config.set('apiVersion', 'v3');
-    RO.config.set('verbose', false);
+    RO.config.init(mockConfig({ apiVersion: 'v3', piiServerUrl: null, verbose: false }));
```
Comment (shamrt) [inline test/resources/v3/orders.test.js:6]: Question: Why is `piiServerUrl: null` declared here?

Comment (shamrt) [reply test/resources/v3/orders.test.js:6]: Ah, because it's set in `mockConfig`. Can we create a `mockPiiConfig`, please, so we have one each for the standard and PII setups? This would decouple these kinds of tests from PII-related tests — and likely eliminate the risk of bugs.

File: test/resources/v3/program-orders.test.js [hunk 13]
```diff
@@ -1,11 +1,14 @@
 const nock = require('nock');
 const RO = require('../../..');
 const fixtures = require('../../fixtures/v3/program-orders.fixtures');
+const { mockConfig } = require('../../test-helpers/mock-config');
+
+RO.config.init(
+  mockConfig({ apiVersion: 'v3', piiServerUrl: null, clientId: 'programTest123', clientSecret: 'itsATestGetUsedToIt' })
```
Comment (shamrt) [inline test/resources/v3/program-orders.test.js:7]: Same

Comment (shamrt) [reply test/resources/v3/program-orders.test.js:7]: Optional: Also, can we leverage the `mockConfig` `clientId` and `clientSecret` props for the — oh, never mind. What a mess. I suppose we need a test helper for creating a Base64 bearer token?

Comment (shamrt) [reply test/resources/v3/program-orders.test.js:7]: Also, to ensure atomicity, `RO.config.init` should go into a `beforeEach` and `afterAll` should convert to a `afterEach`

File: test/resources/v3/program.test.js [hunk 14]
```diff
@@ -1,12 +1,20 @@
...
+
+RO.config.init(
+  mockConfig({
+    apiVersion: 'v3',
+    piiServerUrl: null,
+    clientId: 'programTest123',
+    clientSecret: 'itsATestGetUsedToIt',
+    verbose: false,
+  })
+);
```
Comment (shamrt) [inline test/resources/v3/program.test.js:14]: Same here. Let's at least move it to a `afterEach`

File: test/resources/v3/programs.test.js [hunk 15]
```diff
@@ -1,11 +1,20 @@
 const nock = require('nock');
 const fixtures = require('../../fixtures/v3/programs.fixtures');
 const RO = require('../../..');
+const { mockConfig } = require('../../test-helpers/mock-config');
+
+RO.config.init(
```
Comment (shamrt) [inline test/resources/v3/programs.test.js:6]: Same

File: test/resources/v3/rewards.test.js [hunk 16]
```diff
@@ -1,11 +1,14 @@
 const nock = require('nock');
 const RO = require('../../..');
 const fixtures = require('../../fixtures/v3/rewards.fixtures');
+const { mockConfig } = require('../../test-helpers/mock-config');
+
+RO.config.init(
```
Comment (shamrt) [inline test/resources/v3/rewards.test.js:6]: Same

File: test/resources/v4/custom-categories.test.js [hunk 17]
```diff
@@ -1,11 +1,19 @@
 const nock = require('nock');
 const RO = require('../../..');
 const fixtures = require('../../fixtures/v4/custom-categories.fixtures');
+const { mockConfig } = require('../../test-helpers/mock-config');
+
+RO.config.init(
```
Comment (shamrt) [inline test/resources/v4/custom-categories.test.js:6]: Same. I'll stop now 😄

File: test/resources/v5/order-recipients.test.js [hunk 18]
```diff
@@ -37,7 +42,13 @@ describe('v5 order-recipients', () => {
 
   describe('when piiServerUrl not configured', () => {
     beforeEach(() => {
-      RO.config.set('piiServerUrl', undefined);
+      RO.config.reset();
```
Comment (shamrt) [inline test/resources/v5/order-recipients.test.js:45]: Let's put `RO.config.reset();` into an `afterEach`

Comment (shamrt) [reply test/resources/v5/order-recipients.test.js:45]: And below

File: lib/config.js [hunk 19]
```diff
@@ -63,20 +64,54 @@ const getAll = () => cloneDeep(config);
...
+
+  config = { ...config, [key]: value };
   return value;
 };
 
+let initialized = false;
```
Comment (shamrt) [inline lib/config.js:78]: Let's give this sucker a short, one-liner JSDoc description.

Comment (shamrt) [reply lib/config.js:78]: Optional: Should we also rename it `configInitialized`?

File: lib/config.js [hunk 20]
```diff
@@ -63,20 +64,54 @@ const getAll = () => cloneDeep(config);
...
+
 /**
  * Reset the configuration to defaults
+ *
+ * Reset the `initialized` boolean to false allowing config.init to be called again.
  */
 const reset = () => {
   config = cloneDeep(defaultConfig);
+  initialized = false;
```
Comment (shamrt) [inline lib/config.js:87]: Question: What happens to our auth token if the config is reset? Do we trap ourselves in the same dilemma with standard and PII calls if, for whatever reason, the user sets `piiServerUrl`, resets, then doesn't set that prop again? 

This is likely an extreme edge case. Should we care? should we invalidate the auth token on reset? are there any other edge cases I'm not considering?

File: test/api.test.js [hunk 21]
```diff
@@ -4,10 +4,13 @@ const fixtures = require('./fixtures/api.fixtures');
...
 const emitter = require('../lib/emitter');
+const { mockConfig } = require('../test/test-helpers/mock-config');
+
+const defaultConfig = mockConfig({ apiVersion: 'v4' });
 
 describe('api', () => {
   beforeAll(() => {
-    RO.config.set('apiVersion', 'v4');
+    RO.config.init(defaultConfig);
   });
```
Comment (shamrt) [inline test/api.test.js:14]: These tests need some clean up. They are threateningly un-atomic, which worries me re: bugs. Can we fix them in a follow-up ticket?

File: test/context.test.js [hunk 22]
```diff
@@ -61,7 +62,7 @@ describe('context', () => {
     });
 
     it('should return mockOrderRecipientFactory.create if piiServerUrl is present', () => {
-      config.set('piiServerUrl', faker.internet.url());
+      config.init(mockConfig({ piiServerUrl: faker.internet.url() }));
```
Comment (shamrt) [inline test/context.test.js:65]: Isn't `piiServerUrl: faker.internet.url()` already set in `mockConfig`? This is a tiny example of decoupling troubles, and another basis for using two mock configs, I'd say

**Review by shamrt — CHANGES_REQUESTED** [review]
Some comments 😄

Apologies.

File: test/api.test.js (hunk 21 above)
Comment (Alex-Joseph) [reply test/api.test.js:14]: Created ticket: https://rewardops.atlassian.net/browse/MX-1207

Can you take a look? And add comments about this test file? I am not 100% clear on what you mean by `un-atomic`

Comment (shamrt) [reply test/api.test.js:14]: Here's a decent article on atomic tests: https://techbeacon.com/app-dev-testing/how-make-your-test-automation-run-parallel

File: test/context.test.js (hunk 22 above)
Comment (Alex-Joseph) [reply test/context.test.js:65]: yes, changing `mockConfig({ piiServerUrl: faker.internet.url() })` to `mockConfig()` should not break this test. However, given your previous [comment](https://github.com/rewardops/rewardops-sdk-node/pull/46#discussion_r501051125) I think we should mock `config` so that tests are atomic.

Comment (shamrt) [reply test/context.test.js:65]: > we should mock config so that tests are atomic.

i.e. mock a standard as well as a PII config?

Comment (Alex-Joseph) [reply test/context.test.js:65]: that sounds good

File: lib/config.js (hunk 2 above)
Comment (Alex-Joseph) [reply lib/config.js:72]: 🙏🏻  

yeah sorry, I think I have a pretty elegant solution using the `emitter` module but it's a breaking change. V2 will be better. 

```
try {
  convertToUsingEmitter()
} catch(err) {
  changeTheTopic()
}
```

lol jk

File: lib/schemas/config.js [hunk 23]
```diff
@@ -23,6 +23,7 @@ const configSchema = yup.object().shape({
     .positive()
     .required(),
   verbose: yup.boolean().required(),
+  quiet: yup.boolean().required(),
   supportedLocales: yup.array().nullable(),
```
Comment (shamrt) [inline lib/schemas/config.js:27]: Given this validation assumes the default configs, can we also add a test or 2 in `test/config.test.js` against a `minimalConfig`?

File: lib/schemas/config.js [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+    .default(null),
+  clientId: yup.string().required(),
+  clientSecret: yup.string().required(),
+  logFilePath: yup.string().required(),
+  logToFile: yup.boolean().required(),
+  timeout: yup
+    .number()
+    .positive()
+    .required(),
+  verbose: yup.boolean().required(),
```
[the remark this answers is not in the document]
Comment (Alex-Joseph) [reply lib/schemas/config.js:25]: yes, updated!

File: lib/schemas/config.js (hunk 23 above)
Comment (shamrt) [reply lib/schemas/config.js:27]: I think the only props that are strictly necessary are `clientId` and `clientSecret`

File: lib/config.js (hunk 2 above)
Comment (shamrt) [reply lib/config.js:72]: To confirm: There's a ticket for this?

Comment (Alex-Joseph) [reply lib/config.js:72]: https://rewardops.atlassian.net/browse/MX-1212

File: test/fixtures/v3/program-orders.fixtures.js (hunk 11 above)
Comment (Alex-Joseph) [reply test/fixtures/v3/program-orders.fixtures.js:6]: https://rewardops.atlassian.net/browse/MX-1213

File: lib/config.js [hunk 25]
```diff
@@ -63,20 +64,65 @@ const getAll = () => cloneDeep(config);
...
+ * Boolean that indicates whether the config module has been initialized
+ *
+ * @protected
+ */
+let configInitialized = false;
+
 /**
  * Reset the configuration to defaults
+ *
+ * Reset the `configInitialized` boolean to false allowing config.init to be called again.
```
Comment (shamrt) [inline lib/config.js:88]:
```suggestion
 * Reset the configuration to defaults and allow {@link module:config~init}
 * to be called again.
```

**Review by shamrt — APPROVED** [review]
